src/software/thunderscope/common/proto_configuration_widget.py

```diff
@@ -58,9 +70,199 @@ def __init__(
+    def update_proto_from_file(self, path_to_file: str):
+        """
+        load the protobuf from path_to_file to the variable self.proto_to_configure
+
+        :param path_to_file:
+        """
+        if not os.path.isfile(path_to_file):
+            self.proto_to_configure = ThunderbotsConfig()
+            self.proto_to_configure.sensor_fusion_config.friendly_color_yellow = (
+                self.is_yellow
+            )
+
+            self.build_proto(self.proto_to_configure)
+            self.save_current_config_to_file(self.path_to_file)
+            return
+
+        logging.info("Loading protobuf from file: {}".format(path_to_file))
+        with open(path_to_file, "rb") as f:
+            self.proto_to_configure = ThunderbotsConfig()
+            self.proto_to_configure.ParseFromString(f.read())
+
+            self.proto_to_configure.sensor_fusion_config.friendly_color_yellow = (
+                self.is_yellow
+            )
+
+    def send_proto_to_fullsystem(self):
+        """
+        Sending the default configuration protobufs to unix_full_system at startup.
+        Also updates the widget at the same time.
+        """
+        self.update_proto_from_file(self.path_to_file)
+        self.on_change_callback(None, None, self.proto_to_configure)
+        self.update_widget()
+
+    def save_proto_callback(self):
+        """
+        This is a callback for a button that save the current protobuf to disk!
+        """
+        try:
+            save_to_path, should_save = QFileDialog.getSaveFileName(
+                self,
+                "Select Protobufs",
+                self.get_default_savepath(),
+                options=QFileDialog.Option.DontUseNativeDialog,
+            )
+
+            if not should_save:
+                return
+
+
+            self.save_current_config_to_file(save_to_path)
+            self.update_widget()
+        except Exception:
+            logging.warning(
+                "cannot save configuration to {}".format(
+                    save_to_path
+                )
+            )
+
+    def save_current_config_to_file(self, path_to_file):
+        """
+        This save the self.proto_to_configure to path_to_file
+
+        :param path_to_file: the path to file that we are saving to.
+        """
+
+        logging.info("writing to file {}".format(path_to_file))
+        path_to_directory = os.path.dirname(path_to_file)
+        os.makedirs(path_to_directory, exist_ok=True)
+
+        with open(path_to_file, "wb") as f:
+            f.write(self.proto_to_configure.SerializeToString())
+
+    def load_proto_with_file_explorer(self):
+        """
+        loading the current protobuf through file explorer
+        """
+        try:
+            path_to_file, should_open = QFileDialog.getOpenFileName(
+                self,
+                "Select Protobufs",
+                self.get_default_savepath(),
+                options=QFileDialog.Option.DontUseNativeDialog,
+            )
```

Comment:
I tried saving a custom config as my default, and I was able to relaunch thunderscope with it being automatically loaded and used by fullsystem. I then tried saving a custom config with a custom name, and in subsequent thunderscope launches (defaulting to the first custom config I made), loading the second custom config did not affect fullsystem...

Reply:
That is the intended feature that I am likely going to change. 

You have to overwrite the `default_configuration.proto` with the latest change for it to load. Hence the name `default_configuration.proto`. If you were to save a to a different file, it would not be loaded automatically and you would have to load it manually. 

Do you want me to change this behavior? That is, this widget would load the latest saved proto.

Reply:
I think what you mentioned makes sense and we should load `default_configuration.proto` on startup (assuming it exists). However, the scenario I'm explaining is that I have edited the `default_configuration.proto` to have an obstacle inflation factor of 4. Relaunched Thunderscope, and it loads fine and works as expected. I then modify the obstacle inflation factor to 2, and save the updated config to a new file called `test_configuration.proto`. I relaunch Thunderscope (as expected it loads `default_configuration.proto` with inflation factor=4) and manually load `test_configuration.proto`. The inflation factor is shown to be 2 in the widget, but the actual value shown in Thunderscope and used by full system AI is still 4.